packages/flutter/lib/src/widgets/scrollable_grid.dart

```diff
@@ -0,0 +1,219 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+
+import 'basic.dart';
+import 'framework.dart';
+import 'scrollable.dart';
+
+import 'package:flutter/animation.dart';
+import 'package:flutter/rendering.dart';
+
+/// A vertically scrollable grid.
+///
+/// Requires that delegates place their children in row-major order. Children
+/// can span columns, but the result will not be as efficient as possible.
+class ScrollableGrid extends Scrollable {
+  ScrollableGrid({
+    Key key,
+    double initialScrollOffset,
+    ScrollListener onScroll,
+    SnapOffsetCallback snapOffsetCallback,
+    double snapAlignmentOffset: 0.0,
+    this.delegate,
+    this.children
+  }) : super(
+    key: key,
+    initialScrollOffset: initialScrollOffset,
+    // TODO(abarth): Support horizontal offsets. For horizontally scrollin grids.
+    // For horizontally scrolling grids, we'll probably need to use a delegate
+    // that places children in column-major order.
+    scrollDirection: ScrollDirection.vertical,
+    onScroll: onScroll,
+    snapOffsetCallback: snapOffsetCallback,
+    snapAlignmentOffset: snapAlignmentOffset
+  );
+
+  final GridDelegate delegate;
+  final List<Widget> children;
+
+  ScrollableState createState() => new _ScrollableGrid();
+}
+
+class _ScrollableGrid extends ScrollableState<ScrollableGrid> {
+  ScrollBehavior createScrollBehavior() => new OverscrollWhenScrollableBehavior();
+  ExtentScrollBehavior get scrollBehavior => super.scrollBehavior;
+
+  void _handleExtentsChanged(double contentExtent, double containerExtent) {
+    setState(() {
+      scrollTo(scrollBehavior.updateExtents(
+        contentExtent: contentExtent,
+        containerExtent: containerExtent,
+        scrollOffset: scrollOffset
+      ));
+    });
+  }
+
+  Widget buildContent(BuildContext context) {
+    return new GridViewport(
+      startOffset: scrollOffset,
+      delegate: config.delegate,
+      onExtentsChanged: _handleExtentsChanged,
+      children: config.children
+    );
+  }
+}
+
+typedef void ExtentsChangedCallback(double contentExtent, double containerExtent);
+
+class GridViewport extends RenderObjectWidget {
+  GridViewport({
+    Key key,
+    this.startOffset,
+    this.delegate,
+    this.onExtentsChanged,
+    this.children
+  });
+
+  final double startOffset;
+  final GridDelegate delegate;
+  final ExtentsChangedCallback onExtentsChanged;
+  final List<Widget> children;
+
+  RenderGrid createRenderObject() => new RenderGrid(delegate: delegate);
+
+  _GridViewportElement createElement() => new _GridViewportElement(this);
+}
+
+// TODO(abarth): This function should go somewhere more general.
```

Comment:
I would recommend pooling your algorithms locally and pushing them upstream as you see fit.